src/platforms/hosted/bmp_libusb.c

```diff
@@ -47,291 +106,373 @@ void libusb_exit_function(bmp_info_s *info)
+		dev_info = (FT_DEVICE_LIST_INFO_NODE *)malloc(sizeof(FT_DEVICE_LIST_INFO_NODE) * ftdiDevCount);
+		if (dev_info != NULL) {
+			if (FT_GetDeviceInfoList(dev_info, &ftdiDevCount) == FT_OK) {
+				bool useSerial = true;
+				//
+				// Device list is loaded, iterate over the found probes
+				//
+				for (size_t devIndex = 0; devIndex < ftdiDevCount; devIndex++) {
+					uint16_t vid = (dev_info[devIndex].ID & 0xffff0000) >> 16;
+					uint16_t pid = (dev_info[devIndex].ID & 0xffff);
+					size_t string_len;
+					bool add_probe = true;
 
-	int res = libusb_get_active_config_descriptor(dev, &conf);
-	if (res < 0) {
-		DEBUG_ERROR("libusb_get_active_config_descriptor() failed: %s", libusb_strerror(res));
-		return type;
+					serial = strdup(dev_info[devIndex].SerialNumber);
+					product = strdup(dev_info[devIndex].Description);
+					manufacturer = strdup("FTDI");
+					string_len = strlen(serial);
+					if (!string_len) {
+						free((void *)serial);
+						serial = strdup("---"); // Unknown serial number
+					} else {
+						string_len -= 1;
+						if (*(serial + string_len) == 'A') {
+							*(serial + string_len) = '\0'; // Remove the trailing "A"
+							//
+							// If the serial number is valid, save it for later interface skipping test
+							//
+							if (string_len)
+								probe_skip =
+									strdup(serial); // Save the fixed serial number so we can skip other interfaces
+							string_len = strlen(product);
+							string_len -= 2;                // Product has " A" appended
+							*(product + string_len) = '\0'; // Remove it
+							//
+							// If we don't have a saved serial number, use the truncated product for probe skip test
+							//
+							if (!probe_skip) {
+								useSerial = false;
+								probe_skip = strdup(product);
+							}
+						} else {
+							if (probe_skip) {
+								if (useSerial) {
+									if (strstr(serial, probe_skip))
+										add_probe = false; // Skip this interface
+								} else if (strstr(product, probe_skip))
+									add_probe = false;
+							}
+						}
+					}
+					if (add_probe)
+						probe_list = probe_info_add_by_id(
+							probe_list, BMP_TYPE_FTDI, NULL, vid, pid, manufacturer, product, serial, strdup("---"));
+					else {
+						free(serial);
+						free(product);
+						free(manufacturer);
+					}
+				}
+			}
+			if (probe_skip)
+				free(probe_skip);
+			free((void *)dev_info);
+		} else
+			DEBUG_WARN("process_ftdi_probe: memory allocation failed\n");
 	}
+	return probe_list;
+}
+#endif
 
-	libusb_device_handle *handle;
-	res = libusb_open(dev, &handle);
-	if (res != LIBUSB_SUCCESS) {
-		DEBUG_INFO("libusb_open() failed: %s\n", libusb_strerror(res));
-		libusb_free_config_descriptor(conf);
-		return type;
+void orbtrace_read_version(libusb_device *device, libusb_device_handle *handle, char *version, size_t buffer_size)
+{
+	libusb_config_descriptor_s *config;
+	if (libusb_get_active_config_descriptor(device, &config) == 0) {
```

Comment:
Same comment about pivoting, where possible, to the inverse check and then returning immediately

Reply:
Edit as requested